spec/controllers/sys_controller_spec.rb

```diff
@@ -28,208 +28,368 @@
+        before(:each) do
+          @key = Setting.sys_api_key
+          @project = FactoryGirl.create(:project, :is_public => false)
+          @member = FactoryGirl.create(:member, :user => valid_user,
+                                                :roles => [browse_role],
+                                                :project => @project)
+          request.env['HTTP_AUTHORIZATION'] = ActionController::HttpAuthentication::Basic.encode_credentials(valid_user.login, valid_user_password)
+        end
 
-        random_project = FactoryGirl.create(:project, is_public: false)
-        @member = FactoryGirl.create(:member,
-                                     user: valid_user,
-                                     roles: [browse_role],
-                                     project: random_project)
+        it "should respond 200 okay dokay for read-only access" do
+          post "repo_auth", { :key => @key, :repository => @project.identifier, :method => "GET", :git_smart_http => '1', :uri => '/git', :location => '/git' }
+          expect(response.code).to eq("200")
+        end
 
-        request.env['HTTP_AUTHORIZATION'] = ActionController::HttpAuthentication::Basic.encode_credentials(valid_user.login, valid_user_password)
-        post 'repo_auth', key: @key, repository: @project.identifier, method: 'GET'
+        it "should respond 403 not allowed for write (push)" do
+          post "repo_auth", { :key => @key, :repository => @project.identifier, :method => "POST", :git_smart_http => '1', :uri => "/git/#{@project.identifier}/git-receive-pack", :location => '/git' }
+          expect(response.code).to eq("403")
+        end
       end
 
-      it 'should respond 200 OK' do
-        expect(response.code).to eq('200')
+      context 'for valid login and user has commit access permission (role developer) for project' do
+        before(:each) do
+          @key = Setting.sys_api_key
+          @project = FactoryGirl.create(:project, :is_public => false)
+          @member = FactoryGirl.create(:member, :user => valid_user,
+                                                :roles => [commit_role],
+                                                :project => @project )
+          valid_user.save
+          request.env['HTTP_AUTHORIZATION'] = ActionController::HttpAuthentication::Basic.encode_credentials(valid_user.login, valid_user_password)
+        end
+
+        it "should respond 200 okay dokay for GET" do
+          post "repo_auth", { :key => @key, :repository => @project.identifier, :method => "GET", :git_smart_http => '1', :uri => "/git", :location => '/git' }
+          expect(response.code).to eq("200")
+        end
+
+        it "should respond 200 okay dokay for POST" do
+          post "repo_auth", { :key => @key, :repository => @project.identifier, :method => "POST", :git_smart_http => '1', :uri => "/git/#{@project.identifier}/git-receive-pack", :location => '/git' }
+          expect(response.code).to eq("200")
```

Comment:
Prefer single-quoted strings when you don't need string interpolation or special symbols.